docs/rules/prefer-type-emits-decl.md

```diff
@@ -0,0 +1,52 @@
+---
+pageClass: rule-details
+sidebarDepth: 0
+title: vue/prefer-type-emits-decl
+description: enforce type-based `defineEmits`
+---
+# vue/prefer-type-emits-decl
+
+> enforce type-based `defineEmits`
+
+- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> ***This rule has not been released yet.*** </badge>
+
+## :book: Rule Details
+
+This rule forces developers to use the type-based declaration of `defineEmits` instead of runtime declaration.
+
+This rule only works in setup script and `lang="ts"`.
+
+```vue
+<script setup lang="ts">
+/* ✓ GOOD */
+const emit = defineEmits<{
+  (e: 'change', id: number): void
+  (e: 'update', value: string): void
+}>()
+</script>
+```
+
+<eslint-code-block :rules="{'vue/prefer-type-emits-decl': ['error']}">
+
+```vue
+<script setup lang="ts">
+/* ✗ BAD */
+const emit = defineEmits(['change', 'update'])
+</script>
+```
+
+</eslint-code-block>
+
+## :wrench: Options
+
+Nothing.
```

Comment:
I think it would be great to make the style configurable. I.e.

1. rename rules: `prefer-type-*-decl` → `define-*-declaration`
2. Single string option `type-only` (default) or `runtime-only`

Would you fancy implementing that?

Reply:
Sure, it's a pleasure.

To be consistent with the [Vue documentation](https://vuejs.org/guide/typescript/composition-api.html#typing-component-props), I prefer to name the option as `type-based` and `runtime`.

Reply:
Sure, that names are fine, too :)

As you mention it, could you please add links to these docs pages in the rule docs?

* https://vuejs.org/api/sfc-script-setup.html#defineprops-defineemits
* https://vuejs.org/api/sfc-script-setup.html#typescript-only-features
* https://vuejs.org/guide/typescript/composition-api.html#typing-component-props

Reply:
I think it used to be called "type-only", but thanks to this thread I learned that it has changed to "type-based". 😓 

https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#type-only-propsemit-declarations